[WIP] [Do not merge] Extend keydown input fixes to include modifier keys

cdharris opened the pull request:
Extracted the more complex selection logic out to be functional and added unit tests.

Added controlled text input mechanics for changing a selection with modifier keys

i.e:
	CTRL+Left/Right Arrow = jumps back/forward a word until the start/end
        SHIFT+Left/Right Arrow  = extends the selection back/forward a character until the start/end
	CTRL+SHIFT+Left/Right Arrow = extends the selection back/forward a word until the start/end
	CTRL+Backspace = deletes the previous word

blackforestboi commented:
tested it. Not sure what the state is but here a list of expected behaviours that do not work yet: 
- [ ] Shift + cmd+ arrows > select from cursor to end/beginning of line
- [ ] CMD + backspace > delete line from cursor on
- [ ] Cmd + arrows > jump to beginning of line or end of line/paragraph
- [ ] cursor up and down in text field to jump lines

I also get an error message when building it: 
<img width="604" alt="Screenshot 2019-07-23 at 17 41 42" src="https://user-images.githubusercontent.com/7870039/61726005-28dd1d00-ad71-11e9-9a64-b5634f2e3725.png">

cdharris commented:
Thanks for testing, I've fixed that minor warning.

Those selection modifying key shortcuts you mention are unusal to me, which makes me think they are platform dependent? I take it those are common across all mac software?

I implemented them as I found them in standard Linux (as attached), but it looks like I'll have to check out how they differ across Mac, Linux and Windows.

![keys](https://user-images.githubusercontent.com/3830518/61728594-d3573f00-ad75-11e9-85fb-9ee59da26fb7.gif)

cdharris commented:
Also just included a check for `e.meta` (which seems to be true for cmd on mac) to be synonymous with ctrl for now, so you should at least be able to test the shortcuts as I've implemented them for now.

blackforestboi commented:
with e.meta it works :)

now just the last one is not working correctly then we are good!

blackforestboi commented:
Still a couple of behaviour things that do not work properly: 
- [ ] If I use "cmd+backspace" on a multi-line comment with single words it deletes the whole comment or respectively all lines with just one word that came before. It should only delete the line. It apparently treats multi-line with one word(s) as one line. 
- [ ] If I use "shift + cmd + arrow left" it should select the whole line. Instead it only selects the last word. 

**Prio 2:** 
-[ ] If I open "adding note" with keyboard shortcut c, then use "cmd+enter" it does save the comment, but it does not close the popup like before.

cdharris commented:
Thanks for testing. 

The "cmd+backspace" modifications should do what they are supposed to now, even when lines contain single words or blank lines.

Also, if I understand rightly with the "shift+cmd+left" that it should select word by word up to the whole line, then this is working too, as per the gif above. If those key combinations do other things in other OS', I think that might be out of scope for now?

Had a quick look into automatically closing the popup but didn't turn up any leads. Would like to defer that along with making other usability improvements (like the suggested auto saving). Making time to focus on the Sync work for now.

cdharris commented:
Closing as branch is out of date, will open a new one.
